pulp_2to3_migration/app/tasks/reset.py

```diff
@@ -0,0 +1,73 @@
+import logging
+
+from pulpcore.plugin.models import ProgressReport
+
+from pulp_2to3_migration.app.models import (
+    MigrationPlan,
+    Pulp2Content,
+    Pulp2Distributor,
+    Pulp2Importer,
+    Pulp2LazyCatalog,
+    Pulp2RepoContent,
+    Pulp2Repository,
+)
+
+
+_logger = logging.getLogger(__name__)
+
+
+def reset_pulp3_data(migration_plan_pk):
+    """
+    A task to reset data in Pulp 3 for the plugins specified in a migration plan.
+
+    It also removes all the pre-migrated data. This enables the migration being from scratch and
+    not the incremental one
+
+    Args:
+        migration_plan_pk (str): The migration plan PK.
+
+    """
+    plan = MigrationPlan.objects.get(pk=migration_plan_pk)
+    pulp2_plugins = plan.get_plugins()
+    pb_data = dict(message="Resetting data for Pulp 3 plugins", code="reset.pulp3.data",
+                   total=len(pulp2_plugins))
+    with ProgressReport(**pb_data) as pb:
+        for plugin in plan.get_plugin_plans():
+            for dist_migrator in plugin.migrator.distributor_migrators.values():
+                for dist_model in dist_migrator.pulp3_distribution_models:
+                    dist_model.objects.all().only('pk').delete()
+                for pub_model in dist_migrator.pulp3_publication_models:
+                    pub_model.objects.all().only('pk').delete()
+
+            for imp_migrator in plugin.migrator.importer_migrators.values():
+                for remote_model in imp_migrator.pulp3_remote_models:
+                    remote_model.objects.all().only('pk').delete()
+
+            repo_model = plugin.migrator.pulp3_repository
+            if hasattr(repo_model, 'sub_repo'):
+                # sub_repos can't be deleted until the content referring to them is not removed
+                repo_model.objects.filter(sub_repo=False).only('pk').delete()
+            else:
+                repo_model.objects.all().only('pk').delete()
+
+            for content_model in repo_model.CONTENT_TYPES:
+                content_model.objects.all().only('pk').delete()
```

Comment:
Any suggestions apart from the raw sql to do it differently?

Reply:
It would also be preferable to do it in batches because the memory consumption increases with the amount of content.
For 300 000 rpms, it was 2.6GB. Doesn't seem too bad but still...
To clean up (selectively) data for the large system (1K repos, 300K packages) takes a noticeable time, 30-40 mins.
I'm tempted to write sql, at least for the parts which do not have relations in the tables where data needs to be partially preserved.

Reply:
Unless you explicitly singled out packages in your testing and if this is the same set of repos that contain 600k errata, I would guess that the errata are what takes so long and uses so much memory.  The reason being that Django has to handle the "cascade" deletion, and errata have a lot of object nesting with foreign keys to traverse.

Reply:
600K are errata relations in pulp 2, we count it that way because we need to go through every single one relation and process pkglist for it.
The final number of advisories in pulp 3 is around 35K.
I believe packages are the ones to blame (for memory at least).